### ISSUE TITLE
Questions about Plug and Play Modules

### ISSUE THREAD

**Issue by iWangTing** [MGAMZ/AutoWindow#1]
Dear author, 
﻿ 
First of all, thank you very much for your valuable contributions to this research! 
﻿ 
I have read your paper and part of the related code carefully, but as a beginner, I still have several questions and would like to seek your kind guidance: 
﻿ 
In your experiments, you mentioned using the MedNeXt model based on the nnUNet framework, but in the provided code, it seems that you are using the MMEngine framework instead. Does this mean that you did not use the full nnUNet framework for your experiments, or did you only reuse the MedNeXt model architecture from it? 
﻿ 
In your paper, you proposed a plug-and-play module inspired by the Tanh activation function. Since you implemented your experiments using the MMEngine framework, can this module be used in a standard nnUNet framework in a plug-and-play manner? In other words, is this module independent, or is it tightly coupled with the MMEngine-based network implementation? 
﻿ 
How can I apply the AutoWindow module in a standard nnUNet framework? Are there any key modifications or adaptations needed? 
﻿ 
The module updates the window parameters via backpropagation gradients. Taking U-Net as an example, is the image window updated at the beginning of each epoch, or during every iteration (batch)? 
﻿ 
Thank you again for your time and help! I really appreciate your insights. 
﻿ 
Best regards,

**Comment by MGAMZ** [comment]
Sorry for taking so long to see this message:

1. Yes, we did not conduct experiments based on the `nnUNet` framework, only including its `MedNeXt` neural network design. This is because, at the time, `MMEngine` provided a more flexible foundational training framework. When fairly comparing multiple entirely different neural networks, `MMEngine` offered a better development experience.

2. The proposed method was not designed to depend on the `MMEngine` framework. In our official implementation, however, we adapted it into an MMEngine-compatible form since our experiments were conducted within that framework.

3. Since we haven’t actually implemented this research in `nnUNet`, we cannot confidently provide specific guidance on how to integrate the proposed module into `nnUNet`. However, from a software abstraction perspective, the module is already highly decoupled. Generally speaking, `src.model.ParalleledMultiWindowProcessing` can be plugged into most PyTorch-based frameworks, as it defines a simple `forward` interface and inherits from `torch.nn.Module`. It can be injected as a preprocessing unit before a neural network.

4. In the final version, AutoWindow updates its parameters once per batch. In earlier stages, we experimented with the following learning rate scaling strategy, but it didn’t seem to yield noticeable improvements:

```python
optim_wrapper = dict(
    type=AmpOptimWrapper if use_AMP else OptimWrapper,
    accumulative_counts=grad_accumulation,
    optimizer=dict(type=AdamW,
                   lr=lr,
                   weight_decay=0),
    clip_grad=dict(max_norm=1,
                   norm_type=2,
                   error_if_nonfinite=False),
    paramwise_cfg=dict(
        custom_keys=dict(
            pmwp=dict( # AutoWindow Module has higher LR
                decay_mult=0,
                lr_mult=10))),
)
```

5. Thank you for your interest! Since this work was submitted quite early, if you're planning to reproduce this method using `MMEngine`, you might want to check out another open-source project of mine: https://github.com/MGAMZ/ITKIT. This repository is under active development and includes most of the environment requirements needed to run AutoWindow.

**Comment by iWangTing** [comment]
Thank you so much for your helpful reply!

I really appreciate you taking the time to answer my question. I also checked out the other project you recommended project — it looks fantastic and very well thought out! I’m definitely going to explore it further.

Thanks again for building such great open-source tools! 🙌
